packages/sdk/test/cross-chain-provider.spec.ts

```diff
@@ -1,107 +1,373 @@
 /* eslint-disable @typescript-eslint/no-empty-function */
-import './setup'
+import { expect } from './setup'
+import { Provider } from '@ethersproject/abstract-provider'
+import { Contract } from 'ethers'
+import { ethers } from 'hardhat'
+import {
+  CrossChainProvider,
+  MessageDirection,
+  CONTRACT_ADDRESSES,
+} from '../src'
 
 describe('CrossChainProvider', () => {
   describe('construction', () => {
     describe('when given an ethers provider for the L1 provider', () => {
-      it('should use the provider as the L1 provider', () => {})
+      it('should use the provider as the L1 provider', () => {
+        const provider = new CrossChainProvider({
+          l1Provider: ethers.provider,
+          l2Provider: ethers.provider,
+          l1ChainId: 1,
+        })
+
+        expect(provider.l1Provider).to.equal(ethers.provider)
+      })
     })
 
     describe('when given an ethers provider for the L2 provider', () => {
-      it('should use the provider as the L2 provider', () => {})
+      it('should use the provider as the L2 provider', () => {
+        const provider = new CrossChainProvider({
+          l1Provider: ethers.provider,
+          l2Provider: ethers.provider,
+          l1ChainId: 1,
+        })
+
+        expect(provider.l2Provider).to.equal(ethers.provider)
+      })
     })
 
     describe('when given a string as the L1 provider', () => {
-      it('should create a JSON-RPC provider for the L1 provider', () => {})
+      it('should create a JSON-RPC provider for the L1 provider', () => {
+        const provider = new CrossChainProvider({
+          l1Provider: 'https://localhost:8545',
+          l2Provider: ethers.provider,
+          l1ChainId: 1,
+        })
+
+        expect(Provider.isProvider(provider.l1Provider)).to.be.true
+      })
     })
 
     describe('when given a string as the L2 provider', () => {
-      it('should create a JSON-RPC provider for the L2 provider', () => {})
+      it('should create a JSON-RPC provider for the L2 provider', () => {
+        const provider = new CrossChainProvider({
+          l1Provider: ethers.provider,
+          l2Provider: 'https://localhost:8545',
+          l1ChainId: 1,
+        })
+
+        expect(Provider.isProvider(provider.l2Provider)).to.be.true
+      })
     })
 
     describe('when no custom contract addresses are provided', () => {
       describe('when given a known chain ID', () => {
-        it('should use the contract addresses for the known chain ID', () => {})
+        it('should use the contract addresses for the known chain ID', () => {
+          const provider = new CrossChainProvider({
+            l1Provider: ethers.provider,
+            l2Provider: 'https://localhost:8545',
```

Comment:
just checking that its okay to send a url string for both and not just either/or

Reply:
Yep that's totally fine